applications/upgradeability-by-proxy.md

```diff
@@ -0,0 +1,187 @@
+# Open Grant Proposal
+
+* **Project Name:** ink! Smart Contract Upgradeability
+* **Team Name:** Fractal
+* **Payment Address:** lolcrypto
+
+## Project Overview :page_facing_up:
+
+### Overview
+
+The main goal of this project is to bring smart contract upgradeability to ink.
+
+Smart contract upgradeability is an important mechanism to fix security bugs in
+contracts. It's also important to have flexibility of adding new features.
+
+This mechanism should be flexible enough that contract developers may decide if
+they want to be able to upgrade their contracts, and specify under what
+conditions.
+
+Our team is interested in this because it is very likely that our development
+efforts start with smart contracts before moving to a custom Substrate runtime.
+Upgradeability is important not just because bugs but because product discovery
+will constantly provide new inputs.
+
+
+### Project Details
+
+Upgrade functionality can be implemented in user-land, or as part of the smart
+contracts pallet. We are starting with the first option. This gives us more
+flexibility to experiment and iterate on our solution. When we're happy with
+the result, we'll see what can be migrated to ink as a directive to reduce
+boilerplate.
+
+We're going to start by implementing the proxy pattern. There are a few
```

Comment:
At first I was going to suggest adding something like "taking inspiration from OpenZeppelin ([blog post](https://blog.openzeppelin.com/proxy-patterns/), [docs](https://docs.openzeppelin.com/contracts/3.x/api/proxy)), but then I found this in [the Ethereum docs](https://ethereum.org/en/developers/tutorials/smart-contract-security-guidelines/#upgradeability):

> We discussed the different upgradeability solutions in [our blogpost](https://blog.trailofbits.com/2018/09/05/contract-upgrade-anti-patterns/). Make a deliberate choice to support upgradeability or not prior to writing any code. The decision will influence how you structure our code. In general, we recommend:
> * Favoring [contract migration](https://blog.trailofbits.com/2018/10/29/how-contract-migration-works/) over upgradeability. Migration system have many of the same advantages than upgradeable, without their drawbacks.
> * Using the data separation pattern over the delegatecallproxy one. If your project has a clear abstraction separation, upgradeability using data separation will necessitate only a few adjustments. The delegatecallproxy requires EVM expertise and is highly error-prone.
> * Document the migration/upgrade procedure before the deployment. If you have to react under stress without any guidelines, you will make mistakes. Write the procedure to follow ahead of time. It should include:
>     * The calls that initiate the new contracts
>     * Where are stored the keys and how to access them
>     * How to check the deployment! Develop and test a post-deployment script.

* in the [first linked blog post](https://blog.trailofbits.com/2018/09/05/contract-upgrade-anti-patterns/), they go into more detail of what and why is considered to be an anti-pattern for contract upgrades
* in the [second](https://blog.trailofbits.com/2018/10/29/how-contract-migration-works/), they elaborate further and propose this "contract migration" concept

Can you have a look and see what we can learn from this? That is, if it applies to us at all — I didn't read attentively, but if there's already better established ways to do this, I'd strongly consider revisiting the proposal direction.

Reply:
The "contract migration" proposal is basically "throw out your contract and use a new one, and here's a bunch of issues you'll have doing that, btw". They don't propose a different way of solving upgradeability, they ask you to sidestep it unless you really need it.

The first link has some interesting implementation bugs, some of them are a bit EVM specific, but it's probably good to go through them when implementing the thing.

I don't think there's anything here that influences the proposal, but I do think it's worth keeping these in mind while implementing the mechanism to avoid falling into the same traps.

Reply:
Gotcha, thank you!